sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py

```diff
@@ -94,6 +107,85 @@ def on_request(self, request):
         self._update_headers(request.http_request.headers, self._token.token)
 
 
+class ChallengeAuthenticationPolicy(HTTPPolicy, ABC):
+    """Abstract base class for policies that authorize requests with bearer tokens and expect authentication challenges
+
+    :param ~azure.core.credentials.TokenCredential credential: an object which can provide access tokens, such as a
+        credential from :mod:`azure.identity`
+    :param str scopes: required authentication scopes
+    """
+
+    def __init__(self, credential, *scopes, **kwargs):  # pylint:disable=unused-argument
+        # type: (TokenCredential, *str, **Any) -> None
+        super(ChallengeAuthenticationPolicy, self).__init__()
+        self._scopes = scopes
+        self._credential = credential
+        self._token = None  # type: Optional[AccessToken]
+
+    def _need_new_token(self):
+        # type: () -> bool
+        return not self._token or self._token.expires_on - time.time() < 300
+
+    def authorize_request(self, request, *scopes, **kwargs):
+        # type: (PipelineRequest, *str, **Any) -> None
+        """Acquire a token from the credential and authorize the request with it.
+
+        Keyword arguments are passed to the credential's get_token method. The token will be cached and used to
+        authorize future requests.
+
+        :param ~azure.core.pipeline.PipelineRequest request: the request
+        :param str scopes: required scopes of authentication
+        """
+        self._token = self._credential.get_token(*scopes, **kwargs)
+        request.http_request.headers["Authorization"] = "Bearer " + self._token.token
+
+    def on_request(self, request):
+        # type: (PipelineRequest) -> None
+        """Called before the policy sends a request.
+
+        The base implementation authorizes the request with a bearer token.
+
+        :param ~azure.core.pipeline.PipelineRequest request: the request
+        """
+
+        if self._token is None or self._need_new_token():
+            self._token = self._credential.get_token(*self._scopes)
+        request.http_request.headers["Authorization"] = "Bearer " + self._token.token
+
+    def send(self, request):
+        # type: (PipelineRequest) -> PipelineResponse
+        """Authorizes a request with a bearer token, possibly handling an authentication challenge
+
+        :param ~azure.core.pipeline.PipelineRequest request: the request
+        """
+        _enforce_https(request)
+
+        self.on_request(request)
+
+        response = self.next.send(request)
+
+        if response.http_response.status_code == 401:
+            self._token = None  # any cached token is invalid
+            challenge = response.http_response.headers.get("WWW-Authenticate")
+            if challenge and self.on_challenge(request, response, challenge):
+                response = self.next.send(request)
+
+        return response
+
+    @abc.abstractmethod
+    def on_challenge(self, request, response, challenge):
```

Comment:
By adding the decorator `@abc.abstractmethod` are we forcing subclasses to implement this method even if they don't care to handle challenges? If so should we remove the decorator and make the base simply return false instead?

Reply:
Yes, the decorator mandates an implementation. My thinking is, if you don't want to handle authentication challenges, you want BearerTokenCredentialPolicy instead of ChallengeAuthenticatonPolicy.